### PR TITLE
Avoid double free in distributed tree

### PR DESCRIPTION
**Pull request by dalg24**
Proposed resolution for #368
Stores the duplicated comm in a smart pointer with a custom deleter.

### CODE REVIEW

**Review by aprokop — CHANGES_REQUESTED** [review]
I don't see how it resolves overlapping contexts from multiple copies of the same tree, see [comment](https://github.com/arborx/ArborX/issues/368#issuecomment-691188357).

**Comment by masterleinad** [comment]
Multiple copies would now share the same communicator. What if both of them send/receive messages at the same time? I suspect that would result in potential conflicts.

**Comment by dalg24** [comment]
> I don't see how it resolves overlapping contexts from multiple copies of the same tree, see [comment](https://github.com/arborx/ArborX/issues/368#issuecomment-691188357).

I saw your comment.  This is indeed not addressing the situation you describe but the solution I propose is an improvement over the status quo.  Copies may occur without the user having the intention to search tree in multiple contexts.

**Comment by dalg24** [comment]
From https://github.com/arborx/ArborX/issues/368#issuecomment-691188357
> The fix is straightforward: copy constructor and copy assignment operators must duplicate the communicator.

I was going for compiler generated copy constructor and copy assignment operators but I am fine with the alternate solution you propose.

**Review by aprokop — APPROVED** [review]
I cannot think of a better way to achieve the goal within the scope of this PR.

**Comment by aprokop** [comment]
`#include <shared>`, I guess

File: src/ArborX_DistributedSearchTree.hpp [hunk 1]
```diff
@@ -117,12 +119,11 @@ DistributedSearchTree<MemorySpace, Enable>::DistributedSearchTree(
 
   // Create new context for the library to isolate library's communication from
   // user's
-  auto *newcomm_ptr = new MPI_Comm();
-  MPI_Comm_dup(comm, newcomm_ptr);
-  _comm_ptr.reset(newcomm_ptr, [](MPI_Comm *p) {
+  _comm_ptr.reset(new MPI_Comm(), [](MPI_Comm *p) {
```
Comment (dalg24) [inline src/ArborX_DistributedSearchTree.hpp:122]: I meant to implement it the way I did.  What's the rational for your change?

Comment (dalg24) [reply src/ArborX_DistributedSearchTree.hpp:122]: My rational was if the destructor somehow is called before `MPI_Comm_dup()` is called we cannot recover from it.

Comment (dalg24) [reply src/ArborX_DistributedSearchTree.hpp:122]: I'd be fine with `reset([comm](){ auto p = new MPI_Comm(); MPI_Comm_dup(comm, p); return p; }(), ...)`

Comment (aprokop) [reply src/ArborX_DistributedSearchTree.hpp:122]: I like the latest variant. Please push it.

File: src/ArborX_DistributedSearchTree.hpp [hunk 2]
```diff
@@ -118,12 +119,21 @@ DistributedSearchTree<MemorySpace, Enable>::DistributedSearchTree(
...
+  _comm_ptr.reset(
+      [comm]() {
+        auto p = std::make_unique<MPI_Comm>();
+        MPI_Comm_dup(comm, p.get());
+        return p.release();
+      }(),
+      [](MPI_Comm *p) {
+        MPI_Comm_free(p);
+        delete p;
+      });
```
Comment (aprokop) [inline src/ArborX_DistributedSearchTree.hpp:131]: Please add comments explaining the reasons behind this complication, including what problems it is trying to solve.

File: src/ArborX_DistributedSearchTree.hpp [hunk 3]
```diff
@@ -120,11 +120,14 @@ DistributedSearchTree<MemorySpace, Enable>::DistributedSearchTree(
   // Create new context for the library to isolate library's communication from
   // user's
   _comm_ptr.reset(
+      // duplicate the communicator and store it in a std::shared_ptr so that
+      // all copies of the distributed tree point to the same object
       [comm]() {
         auto p = std::make_unique<MPI_Comm>();
```
Comment (aprokop) [inline src/ArborX_DistributedSearchTree.hpp:126]: I actually wanted to see a comment on why `unique_ptr` is necessary here.

Comment (dalg24) [reply src/ArborX_DistributedSearchTree.hpp:126]: Manages lifetime of the pointer until ownership is transferred to the `std::shared_ptr` (would call `delete` if anything goes wrong and an exception is thrown before the `shared_ptr` is constructed)

Comment (aprokop) [reply src/ArborX_DistributedSearchTree.hpp:126]: From conversation with @dalg24, the point is that if we create a pointer and then something happens inside `MPI_Comm_dup()` in that lambda, the pointer is never deleted. Wrapping it in the unique pointer takes care of that.

**Review by aprokop — APPROVED** [review]

File: src/ArborX_DistributedSearchTree.hpp (hunk 3 above)
Comment (masterleinad) [reply src/ArborX_DistributedSearchTree.hpp:126]: I don't feel strongly enough about this whole issue to argue against anything the two of you have agreed upon. If we are that concerned about errors in MPI calls, checking the return values feels appropriate, though.

Comment (dalg24) [reply src/ArborX_DistributedSearchTree.hpp:126]: >  If we are that concerned about errors in MPI calls, checking the return values feels appropriate, though.

This is a larger issue.  The user may have specified an error handler that throws.

**Comment by dalg24** [comment]
Retest this please

**Comment by dalg24** [comment]
Do we really care to wait for the HIP build to be fixed?

**Comment by masterleinad** [comment]
Blame me if there are any warnings we were missing.

**Pull request merged**